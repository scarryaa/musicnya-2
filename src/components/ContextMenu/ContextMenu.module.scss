:root {
  --contextmenu-background: rgba(204, 204, 204, 0.8);
  --contextmenu-color: #000;
}

:root[theme='dark'] {
  --contextmenu-background: rgba(39, 39, 39, 0.8);
  --contextmenu-color: #fff;
}

.contextMenu {
  position: absolute;
  z-index: 100000;
  background: var(--contextmenu-background);
  border-radius: 4px;
  padding: 0.5rem;
  color: var(--contextmenu-color);
  box-shadow: 0 0 20px rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(10px);
  min-width: 150px;
}

.contextMenu__subContextMenu {
  position: absolute;
  z-index: 100;
  background: var(--contextmenu-background);
  border-radius: 4px;
  padding: 0.5rem 0;
  color: var(--contextmenu-color);
  box-shadow: 0 0 20px rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(10px);
  overflow-y: auto;
  height: 200px;
  top: 0;
  left: 100%;
  width: 100%;
}

.contextMenu__quick__item {
  padding: 0.4rem;
  border-radius: 4px;

  &:hover {
    background: rgba(255, 255, 255, 0.15);
  }
}

.contextMenu__item {
  padding: 0.2rem 0.5rem;
  font-size: 0.875rem;
  display: flex;
  align-items: center;
  gap: 0.5rem;

  &:hover {
    background: rgba(255, 255, 255, 0.15);
  }

  &__icon {
    display: flex;
    align-items: center;
  }

  &__subMenuIcon {
    margin-left: auto;
  }

  &__label {
    vertical-align: top;
    display: inline-block;
    margin-top: 6px;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
  }
}

.quickActions {
  display: flex;
  flex-direction: row;
}
