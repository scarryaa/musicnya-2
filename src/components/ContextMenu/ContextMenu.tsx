import { For, JSX, createEffect, onCleanup, onMount } from 'solid-js'
import styles from './ContextMenu.module.scss'
import { store } from '../../stores/store'
import Fa from 'solid-fa'
import { faChevronRight } from '@fortawesome/free-solid-svg-icons'
import { useContextMenu, useContextMenuState } from '../../composables/useContextMenu'
import { ContextMenuType, Reaction } from '../../types/types'
import Tooltip from '../Tooltip/Tooltip'
import { historyItemContextMenuConfig } from './Configs/historyItemContextMenuConfig'
import { appContextMenuConfig } from './Configs/appContextMenuConfig'
import { artistContextMenuConfig } from './Configs/artistContextMenuConfig'
import { songContextMenuConfig } from './Configs/songContextMenuConfig'
import { mediaItemContextMenuConfig } from './Configs/mediaItemContextMenuConfig'
import { queueItemContextMenuConfig } from './Configs/queueItemContextMenuConfig'
import { editorialContextMenuConfig } from './Configs/editorialContextMenuConfig'
import { QuickActionItem } from './Components/QuickActionItem'

const contextMenuConfig = {
  [ContextMenuType.App]: appContextMenuConfig,
  [ContextMenuType.HistoryItem]: historyItemContextMenuConfig,
  [ContextMenuType.Artist]: artistContextMenuConfig,
  [ContextMenuType.Song]: songContextMenuConfig,
  [ContextMenuType.MediaItem]: mediaItemContextMenuConfig,
  [ContextMenuType.QueueItem]: queueItemContextMenuConfig,
  [ContextMenuType.Editorial]: editorialContextMenuConfig
}

//TODO fix editorial item small context menu
export function ContextMenu(): JSX.Element {
  const { setContextMenuItems } = useContextMenu()
  const { closeContextMenu } = useContextMenuState()

  const updateMenuItems = async (type, subType) => {
    if (!contextMenuConfig[type]) {
      console.error('Unsupported context menu type')
      return
    }

    const data = await contextMenuConfig[type].fetchData(
      store.app.contextMenu.id,
      subType
    )
    const updatedItems = contextMenuConfig[type].createMenuItems(
      store.app.contextMenu.id,
      subType,
      data
    )
    setContextMenuItems(updatedItems)
  }

  onMount(() => {
    window.addEventListener('click', closeContextMenu)
    onCleanup(() => window.removeEventListener('click', closeContextMenu))
  })

  createEffect(() => {
    const menuType = store.app.contextMenu.type
    if (!contextMenuConfig[menuType]) {
      return
    }

    const initialMenuItems = contextMenuConfig[menuType].createInitialMenuItems(
      store.app.contextMenu.id,
      store.app.contextMenu.subType
    )
    setContextMenuItems(initialMenuItems)

    if (menuType !== ContextMenuType.Curator && menuType !== ContextMenuType.App) {
      updateMenuItems(menuType, store.app.contextMenu.subType)
    }
  }, [store.app.contextMenu.id, store.app.contextMenu.type])

  const filterMenuItems = isQuickAction =>
    store.app.contextMenu.items.filter(
      item => item && item.isQuickAction === isQuickAction
    )

  return (
    <div
      class={styles.contextMenu}
      style={`top: ${store.app.contextMenu.y}px; left: ${store.app.contextMenu.x}px; display: ${store.app.contextMenu.display}`}
    >
      <div class={styles.quickActions}>
        <For each={filterMenuItems(true)}>{item => <QuickActionItem item={item} />}</For>
      </div>

      <For each={filterMenuItems(false)}>
        {item => (
          <div
            class={styles.contextMenu__item}
            onclick={item.action}
            onMouseOver={item.onMouseOver}
            onMouseLeave={item.onMouseLeave}
          >
            <Fa icon={item.icon} class={styles.icon} />
            <span class={styles.label}>{item.label}</span>
            {item.hasSubMenu && <Fa icon={faChevronRight} class={styles.subMenuIcon} />}
          </div>
        )}
      </For>

      <div
        class={styles.contextMenu__subContextMenu}
        style={`display: ${
          store.app.subContextMenu.open ? 'block' : 'none'
        }; top: calc(0px + ${
          store.app.subContextMenu.y
        }px); left: 100%; position: absolute; width: 100%; max-height: 200%; overflow-y: auto;`}
      >
        <For each={store.app.subContextMenu.items}>
          {item => (
            <div
              class={styles.contextMenu__item}
              onclick={item.hasSubMenu ? null : item.action}
              onMouseOver={item.onMouseOver}
              onMouseLeave={item.onMouseLeave}
            >
              <span class={styles.contextMenu__item__icon}>
                <Fa icon={item.icon} size="1x" color="white" />
              </span>
              <span
                style={`vertical-align: top; margin-left: 0.5rem; margin-top: 6px; text-overflow: ellipsis; overflow: hidden;`}
              >
                {item.label}
              </span>
            </div>
          )}
        </For>
      </div>
    </div>
  )
}
