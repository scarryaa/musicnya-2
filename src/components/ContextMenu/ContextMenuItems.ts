import {
  IconDefinition,
  faHeadphones,
  faHeart,
  faIndent,
  faInfoCircle,
  faMinus,
  faOutdent,
  faPlus,
  faRecordVinyl,
  faSatelliteDish,
  faShare,
  faShuffle,
  faStar,
  faThumbsDown,
  faUpRightAndDownLeftFromCenter,
  faUser,
  faX
} from '@fortawesome/free-solid-svg-icons'
import {
  faHeart as faHeartRegular,
  faThumbsDown as faThumbsDownRegular,
  faStar as faStarRegular,
  faWindowMaximize
} from '@fortawesome/free-regular-svg-icons'
import { setStore, store } from '../../stores/store'
import { Utils } from '../../util/util'
import { usePlaylistSubMenu } from '../../composables/usePlaylistSubMenu'
import { useSubContextMenuState } from '../../composables/useSubContextMenu'
import { mkManager } from '../../api/MkManager'
import { mkApiManager } from '../../api/MkApiManager'

const { closeSubContextMenu } = useSubContextMenuState()

const createMenuItem = (
  icon: IconDefinition,
  action: () => Promise<void> | void,
  label: string,
  quickAction = false,
  hasSubMenu = false,
  tooltip?: string
) => {
  return {
    icon,
    action,
    isQuickAction: quickAction,
    label,
    hasSubMenu,
    onMouseEnter: () => closeSubContextMenu(),
    tooltip
  }
}

const createAsyncMenuItem = (
  icon: IconDefinition,
  asyncAction: () => Promise<void>,
  label: string,
  quickAction = false,
  tooltip?: string
) => {
  return {
    icon,
    action: async () => {
      try {
        await asyncAction()
      } catch (error) {
        console.error('Error executing menu action:', error)
      }
    },
    isQuickAction: quickAction,
    label,
    onMouseEnter: () => closeSubContextMenu(),
    tooltip
  }
}

// Actions
export const contextMenuItems = {
  immersive: createMenuItem(faUpRightAndDownLeftFromCenter, () => {}, 'Immersive'),
  miniPlayer: createMenuItem(faWindowMaximize, Utils.resizeToMiniPlayer, 'Mini Player'),
  share: (id: string, type: MusicKit.MediaItemType) =>
    createAsyncMenuItem(
      faShare,
      async () => {
        const data = await mkApiManager.getShareLink(id, type)
        await navigator.clipboard.writeText(data.data[0].attributes.url)
      },
      'Share'
    ),
  properties: (id: string, type: MusicKit.MediaItemType) =>
    createMenuItem(
      faInfoCircle,
      () => {
        setStore('app', 'modal', { open: true, type, id })
      },
      'Properties'
    ),
  createStation: (id: string, type: MusicKit.MediaItemType) =>
    createMenuItem(
      faSatelliteDish,
      async () => {
        await mkManager.setStationQueue(id, type)
      },
      'Create Station'
    ),
  goToArtist: (id: string, type: MusicKit.MediaItemType) =>
    createAsyncMenuItem(
      faUser,
      async () => {
        const data = await mkApiManager.getArtistFromMediaItem(id, type)
        console.log(data)
        store.app.navigate(`/media/artists/${data.data[0].id}`)
      },
      'Go to Artist'
    ),

  goToAlbum: (id: string, type: MusicKit.MediaItemType) =>
    createAsyncMenuItem(
      faRecordVinyl,
      async () => {
        const data = await mkApiManager.getAlbumFromMediaItem(id, type)
        console.log(data)
        store.app.navigate(`/media/albums/${data.data[0].id}`)
      },
      'Go to Album'
    ),

  shuffle: (id: string, type: MusicKit.MediaItemType) =>
    createMenuItem(
      faShuffle,
      async () => mkManager.processItemAndPlay(id, type, true),
      'Shuffle'
    ),

  removeFromQueue: (id: string) =>
    createMenuItem(
      faX,
      () => {
        const returnedId = mkManager.removeFromQueue(id)
        setStore('app', 'queue', {
          items: store.app.queue.items.filter(
            (item: MusicKit.MediaItem) => item.id !== returnedId
          )
        })
      },
      'Remove from Queue'
    ),

  addToPlaylist: (id, type) => ({
    ...createMenuItem(faHeadphones, () => {}, 'Add to Playlist', false, true),
    onMouseEnter: e => {
      const { openPlaylistSubMenu } = usePlaylistSubMenu(e)
      openPlaylistSubMenu(id, type)
    }
  }),
  // Quick actions
  favoriteQuick: (id: string, type, disabled: boolean, isFavorite = false) =>
    createMenuItem(
      disabled ? faStarRegular : isFavorite ? faStar : faStarRegular,
      () =>
        isFavorite ? mkApiManager.unfavoriteArtist(id) : mkApiManager.favoriteArtist(id),
      isFavorite ? 'Unfavorite' : 'Favorite',
      true,
      false,
      isFavorite ? 'Unfavorite' : 'Favorite'
    ),

  createStationQuick: (id: string, type: MusicKit.MediaItemType) =>
    createMenuItem(
      faSatelliteDish,
      () => mkManager.setStationQueue(id, type),
      'Create Station',
      true,
      false,
      'Create Station'
    ),

  loveQuick: (
    id: string,
    type: MusicKit.MediaItemType,
    disabled: boolean,
    isLoved = false
  ) =>
    createMenuItem(
      disabled ? faHeartRegular : isLoved ? faHeart : faHeartRegular,
      () =>
        isLoved
          ? mkApiManager.unfavoriteItem(id, type)
          : mkApiManager.favoriteItem(id, type),
      isLoved ? 'Unlove' : 'Love',
      true,
      false,
      isLoved ? 'Unlove' : 'Love'
    ),

  dislikeQuick: (
    id: string,
    type: MusicKit.MediaItemType,
    disabled: boolean,
    isLoved: boolean,
    isDisliked = false
  ) =>
    createMenuItem(
      disabled ? faThumbsDownRegular : isDisliked ? faThumbsDown : faThumbsDownRegular,
      () =>
        isDisliked
          ? mkApiManager.unfavoriteItem(id, type)
          : mkApiManager.dislikeItem(id, type),
      isDisliked ? 'Undislike' : 'Dislike',
      true,
      false,
      isDisliked ? 'Undislike' : 'Dislike'
    ),

  addToLibraryQuick: (
    id: string,
    type: MusicKit.MediaItemType,
    disabled,
    isLoved,
    isDisliked,
    isInLibrary = false
  ) =>
    createAsyncMenuItem(
      disabled ? faPlus : isInLibrary ? faMinus : faPlus,
      async () => {
        if (isInLibrary && id[0] !== 'l') {
          const resp = await mkApiManager.getLibraryIdFromCatalog(id, type)
          await mkApiManager.removeFromLibrary(resp.data[0].id, type)
        } else if (isInLibrary) {
          await mkApiManager.removeFromLibrary(id, type)
        } else {
          await mkApiManager.addToLibrary(id, type)
        }
      },
      isInLibrary ? 'Remove from Library' : 'Add to Library',
      true,
      isInLibrary ? 'Remove from Library' : 'Add to Library'
    ),

  playNextQuick: (id: string, type: MusicKit.MediaItemType) =>
    createMenuItem(
      faIndent,
      () => mkManager.playNext(id, type),
      'Play Next',
      true,
      false,
      'Play Next'
    ),

  playLastQuick: (id: string, type: MusicKit.MediaItemType) =>
    createMenuItem(
      faOutdent,
      () => mkManager.playLater(id, type),
      'Play Last',
      true,
      false,
      'Play Last'
    )
}
