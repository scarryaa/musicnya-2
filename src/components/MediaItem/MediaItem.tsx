import styles from './MediaItem.module.scss'
import { A } from '@solidjs/router'
import Tooltip from '../Tooltip/Tooltip'
import { useContextMenu } from '../../composables/useContextMenu'
import { ContextMenuType, MediaItemType } from '../../types/types'
import useMediaItem from '../../composables/useMediaItem'
import { ArtworkOverlay, ArtworkOverlayType } from '../ArtworkOverlay/ArtworkOverlay'

export const MediaItem = ({
  src,
  title,
  artists,
  type,
  id,
  artistId,
  releaseYear,
  curator,
  curatorId
}: {
  src: string
  title?: string
  artists: string[]
  type: MediaItemType
  id: string
  artistId: string
  releaseYear?: number
  curator?: string
  curatorId?: string
}) => {
  Tooltip
  const { openContextMenu } = useContextMenu()
  const { newArtistId, handlePlayClick } = useMediaItem(type, id, artistId)

  const isAppleCurator = type === 'apple-curators'
  type === 'library-playlists' ? (type = MediaItemType.Playlists) : type

  return (
    <div
      class={styles['media-item']}
      data-id={id}
      data-testid="media-item"
      onContextMenu={e => openContextMenu(e, id, ContextMenuType.MediaItem, type)}
    >
      <div class={styles['media-item__inner']}>
        <div class={styles['media-item__inner__artwork']}>
          <ArtworkOverlay
            type={
              isAppleCurator ? ArtworkOverlayType.MORE : ArtworkOverlayType.PLAY_AND_MORE
            }
            isLink={true}
            link={isAppleCurator ? `/media/curators/${id}` : `/media/${type}/${id}`}
            playClick={handlePlayClick}
            moreClick={e => openContextMenu(e, id, ContextMenuType.MediaItem, type)}
          >
            <img
              class={styles['media-item__inner__artwork__image']}
              src={src}
              alt={title}
            />
          </ArtworkOverlay>
        </div>
        <div class={styles['media-item__inner__info']}>
          <div
            class={styles['media-item__inner__info__title']}
            use:Tooltip={['bottom', title, true, 500]}
          >
            {title}
          </div>
          <div class={styles['media-item__inner__info__artists']}>
            {releaseYear && (
              <span class={styles['media-item__inner__info__artists__year']}>
                {releaseYear}
              </span>
            )}
            {artists !== null && (
              <A
                activeClass=""
                data-testid="artist-link"
                class={styles['media-item__inner__info__artists__link']}
                href={`/media/artists/${newArtistId()}`}
              >
                <span use:Tooltip={['bottom', artists.join(', '), true, 500]}>
                  {artists.join(', ')}
                </span>
              </A>
            )}
            {curator &&
              (curatorId ? (
                <A
                  activeClass=""
                  class={styles['media-item__inner__info__curator-link']}
                  href={`/media/artists/${curatorId}`}
                >
                  <span use:Tooltip={['bottom', curator, true, 500]}>{curator}</span>
                </A>
              ) : (
                <span
                  use:Tooltip={['bottom', curator, true, 500]}
                  class={styles['media-item__inner__info__curator']}
                >
                  {curator}
                </span>
              ))}
          </div>
        </div>
      </div>
    </div>
  )
}
