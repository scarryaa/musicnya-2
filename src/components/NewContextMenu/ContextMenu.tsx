import { createEffect, createSignal, onCleanup, onMount } from 'solid-js'
import useNewContextMenu from '../../composables/useNewContextMenu'
import { ContextMenuType } from '../../types/types'
import { ContextMenuItems } from './Components/ContextMenuItems/ContextMenuItems'
import { ContextMenuQuickItems } from './Components/ContextMenuQuickItems/ContextMenuQuickItems'
import { contextMenuConfig } from './Configs/mainConfig'
import styles from './ContextMenu.module.scss'
import { updateMenuItems } from './Helpers/ContextMenuHelpers'
import { newContextMenuStore } from '../../stores/newContextMenuStore'
import { SubContextMenu } from './Components/SubContextMenu/SubContextMenu'

export const ContextMenu = () => {
  const [tabbedFromMenuItem, setTabbedFromMenuItem] = createSignal(null as Element | null)
  const { open, menuPosition, closeNewContextMenu, setContextMenuItems } =
    useNewContextMenu()

  const handleKeyDown = (e: KeyboardEvent) => {
    // close the context menu if the user presses escape
    if (e.key === 'Escape') {
      closeNewContextMenu()
    } else if (e.key === 'ArrowDown') {
      // cycle through the menu items, looping back to the top if the user reaches the bottom
      const contextMenu = document.getElementById('context-menu')
      const contextMenuItems = contextMenu?.querySelectorAll('li')
      const activeElement = document.activeElement
      const activeElementIndex = Array.from(contextMenuItems!).indexOf(activeElement)
      const nextElement = contextMenuItems?.[activeElementIndex + 1] as HTMLElement
      nextElement?.focus()
    } else if (e.key === 'ArrowUp') {
      // cycle through the menu items, looping back to the bottom if the user reaches the top
      const contextMenu = document.getElementById('context-menu')
      const contextMenuItems = contextMenu?.querySelectorAll('li')
      const activeElement = document.activeElement
      const activeElementIndex = Array.from(contextMenuItems!).indexOf(activeElement)
      const nextElement = contextMenuItems?.[activeElementIndex - 1] as HTMLElement
      nextElement?.focus()
    } else if (e.key === 'ArrowRight') {
      e.preventDefault()
      // set the tabbed from menu item to the current menu item
      setTabbedFromMenuItem(document.activeElement)

      // open the sub context menu if the user presses the right arrow key
      const subContextMenu = document.getElementById('subContextMenu')
      const subContextMenuItems = subContextMenu?.querySelectorAll('li')
      subContextMenuItems?.[0].focus()
    } else if (e.key === 'ArrowLeft') {
      e.preventDefault()
      console.log(tabbedFromMenuItem)
      // return to the tabbed from menu item if the user presses the left arrow key
      tabbedFromMenuItem()?.focus()
    }
  }

  onMount(() => {
    window.addEventListener('click', closeNewContextMenu)
  })

  onCleanup(() => {
    window.removeEventListener('click', closeNewContextMenu)
  })

  createEffect(() => {
    const menuType = newContextMenuStore.type

    if (
      !newContextMenuStore.open ||
      !newContextMenuStore.type ||
      menuType === ContextMenuType.View
    ) {
      return
    }

    const initialMenuItems = contextMenuConfig[menuType].createInitialMenuItems(
      newContextMenuStore.id,
      newContextMenuStore.subtype
    )

    setContextMenuItems(initialMenuItems)

    if (menuType !== ContextMenuType.Curator && menuType !== ContextMenuType.App) {
      updateMenuItems(
        menuType,
        newContextMenuStore.subtype as MusicKit.MediaItemType,
        contextMenuConfig,
        setContextMenuItems
      ).catch(err => console.error(err))
    }

    // focus the menu when it opens
    setTimeout(() => {
      document.getElementById('context-menu')?.focus()
    })
  }, [newContextMenuStore.id, newContextMenuStore.type])

  console.log('ContextMenuItems', newContextMenuStore.items)
  return (
    <div
      tabindex={-1}
      onKeyDown={(e: KeyboardEvent) => handleKeyDown(e)}
      role="menubar"
      id="context-menu"
      class={styles['context-menu']}
      style={`top: ${menuPosition().y}px; left: ${menuPosition().x}px; display: ${
        open() ? 'block' : 'none'
      }`}
    >
      <ContextMenuQuickItems />
      <ContextMenuItems />
      <SubContextMenu />
    </div>
  )
}
