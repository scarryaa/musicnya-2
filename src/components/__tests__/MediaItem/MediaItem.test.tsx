jest.mock('../../../stores/store', () => ({
  setStore: jest.fn(),
  store: {
    app: {
      contextMenu: {
        open: false,
        x: 0,
        y: 0,
        id: '',
        type: null,
        items: [],
        subType: null,
        display: 'none'
      },
      subContextMenu: {
        x: 0,
        y: 0,
        items: [],
        open: false,
        id: '',
        type: ''
      }
    }
  }
}))

jest.mock('../../../composables/useContextMenu')
jest.mock('../../../api/MkApiManager.ts')
jest.mock('../../../api/MkManager.ts')

useContextMenu.mockReturnValue({
  openContextMenu: jest.fn()
})

useContextMenuState.mockReturnValue({
  closeContextMenu: jest.fn()
})

mkApiManager.getCatalogArtistFromLibrary.mockResolvedValue({
  data: [
    {
      id: '1'
    }
  ]
})

mkManager.processItemAndPlay.mockResolvedValue({
  data: [
    {
      id: '1'
    }
  ]
})

import { MediaItem } from '../../MediaItem/MediaItem'
import '@testing-library/jest-dom'
import { render, screen, fireEvent, waitFor } from '@solidjs/testing-library'
import { Router } from '@solidjs/router'
import { ContextMenu } from '../../ContextMenu/ContextMenu'
import { useContextMenu, useContextMenuState } from '../../../composables/useContextMenu'
import { mkApiManager } from '../../../api/MkApiManager'
import { mkManager } from '../../../api/MkManager'
import { MediaItemType } from '../../../types/types'

describe('MediaItem', () => {
  test('renders MediaItem component correctly', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    expect(screen.getByTestId('play-button')).toBeInTheDocument()
    expect(screen.getByText('Artist 1')).toBeInTheDocument()
  })

  test('does not render play button when type is AppleCurators', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.AppleCurators}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    const playButton = screen.queryByTestId('play-button')
    expect(playButton).not.toBeInTheDocument()
    expect(screen.getByText('Artist 1')).toBeInTheDocument()
  })

  test('renders MediaItem component correctly when curator and curatorId are present', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
          curator="Curator 1"
          curatorId="1"
        />
      </Router>
    ))
    expect(screen.getByText('Curator 1')).toBeInTheDocument()
  })

  test('renders MediaItem component correctly when releaseYear is present', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
          releaseYear={2021}
        />
      </Router>
    ))
    expect(screen.getByText('2021')).toBeInTheDocument()
  })

  test('renders MediaItem component correctly when releaseYear is not present', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    expect(screen.queryByText('2021')).not.toBeInTheDocument()
  })

  test('renders MediaItem component correctly when artistId is present', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.LibraryAlbums}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    expect(screen.getByText('Artist 1')).toBeInTheDocument()
  })

  test('renders MediaItem component correctly when optional props are missing', () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    expect(screen.getByText('Artist 1')).toBeInTheDocument()
  })

  test('renders MediaItem component correctly when artists is missing', () => {
    render(() => (
      <Router>
        <MediaItem
          artists={null}
          src="test.jpg"
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    const artistLink = screen.queryByTestId('artist-link')
    expect(artistLink).not.toBeInTheDocument()
  })

  test('right click on media item opens context menu', async () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
        <ContextMenu />
      </Router>
    ))

    const mediaItem = screen.getByTestId('media-item')
    fireEvent.contextMenu(mediaItem)

    const { openContextMenu } = useContextMenu()

    await waitFor(() => {
      expect(openContextMenu).toHaveBeenCalledWith(
        expect.anything(),
        '1',
        'mediaItem',
        MediaItemType.Albums
      )
    })
  })

  test('click on more button opens context menu', async () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
        <ContextMenu />
      </Router>
    ))

    const moreButton = screen.getByTestId('more-button')
    fireEvent.click(moreButton)

    const { openContextMenu } = useContextMenu()

    await waitFor(() => {
      expect(openContextMenu).toHaveBeenCalledWith(
        expect.anything(),
        '1',
        'mediaItem',
        MediaItemType.Albums
      )
    })
  })

  test('click on play button calls processItemAndPlay', async () => {
    render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
        <ContextMenu />
      </Router>
    ))

    const playButton = screen.getByTestId('play-button')
    fireEvent.click(playButton)

    await waitFor(() => {
      expect(mkManager.processItemAndPlay).toHaveBeenCalledWith('1', 'albums')
    })
  })
})

describe('MediaItem Snapshot', () => {
  it('should match snapshot', () => {
    const { container } = render(() => (
      <Router>
        <MediaItem
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
          id="1"
          artistId="1"
        />
      </Router>
    ))
    expect(container.firstChild).toMatchSnapshot()
  })
})

describe('MediaItem Responsive Tests', () => {
  const originalInnerWidth = global.innerWidth

  beforeEach(() => {
    // Reset window dimensions after each test
    global.innerWidth = originalInnerWidth
    global.dispatchEvent(new Event('resize'))
  })

  it('renders correctly on mobile viewports', () => {
    global.innerWidth = 375 // Mobile width
    global.dispatchEvent(new Event('resize'))

    render(() => (
      <Router>
        <MediaItem
          artistId="1"
          id="1"
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
        />
      </Router>
    ))

    expect(screen.getByTestId('play-button')).toBeInTheDocument()
    expect(screen.getByText('Artist 1')).toBeInTheDocument()
  })

  it('renders correctly on tablet viewports', () => {
    global.innerWidth = 768 // Tablet width
    global.dispatchEvent(new Event('resize'))

    render(() => (
      <Router>
        <MediaItem
          artistId="1"
          id="1"
          src="test.jpg"
          artists={['Artist 1']}
          type={MediaItemType.Albums}
        />
      </Router>
    ))

    expect(screen.getByTestId('play-button')).toBeInTheDocument()
    expect(screen.getByText('Artist 1')).toBeInTheDocument()
  })
})
