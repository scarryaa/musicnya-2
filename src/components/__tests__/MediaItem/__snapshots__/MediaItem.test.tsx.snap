// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaItem Snapshot should match snapshot 1`] = `
<div
  class="media-item"
  data-id="1"
  data-testid="media-item"
>
  <div
    class="media-item__inner"
  >
    <div
      class="media-item__inner__artwork"
    >
      <div
        class="artwork-overlay"
      >
        <a
          aria-current="page"
          class="artwork-overlay-link"
          data-testid="artwork-overlay-link"
          href="/media/albums/1"
          link=""
        >
          <div
            class="artwork-overlay-link-play-button"
            data-testid="play-button"
          >
            <svg
              aria-hidden="true"
              class="solid-fa"
              style="height: 1em; font-size: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(192 256)"
                transform-origin="96 0"
              >
                <g
                  transform="translate(0,0) scale(1,1)"
                >
                  <path
                    d="M73 39c-14.8-9.1-33.4-9.4-48.5-.9S0 62.6 0 80V432c0 17.4 9.4 33.4 24.5 41.9s33.7 8.1 48.5-.9L361 297c14.3-8.7 23-24.2 23-41s-8.7-32.2-23-41L73 39z"
                    fill="var(--color-white)"
                    transform="translate(-192 -256)"
                  />
                </g>
              </g>
            </svg>
          </div>
          <div
            class="artwork-overlay-link-more-button"
            data-testid="more-button"
          >
            <svg
              aria-hidden="true"
              class="solid-fa"
              style="height: 1em; font-size: 1em; vertical-align: -.125em; transform-origin: center; overflow: visible;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                transform="translate(224 256)"
                transform-origin="112 0"
              >
                <g
                  transform="translate(0,0) scale(1,1)"
                >
                  <path
                    d="M8 256a56 56 0 1 1 112 0A56 56 0 1 1 8 256zm160 0a56 56 0 1 1 112 0 56 56 0 1 1 -112 0zm216-56a56 56 0 1 1 0 112 56 56 0 1 1 0-112z"
                    fill="var(--color-white)"
                    transform="translate(-224 -256)"
                  />
                </g>
              </g>
            </svg>
          </div>
        </a>
        <img
          class="media-item__inner__artwork__image"
          src="test.jpg"
        />
      </div>
    </div>
    <div
      class="media-item__inner__info"
    >
      <div
        class="media-item__inner__info__title"
      />
      <div
        class="media-item__inner__info__artists"
      >
        
        <a
          class="media-item__inner__info__artists__link inactive"
          data-testid="artist-link"
          href="/media/artists/1"
          link=""
        >
          <span>
            Artist 1
          </span>
        </a>
        
      </div>
    </div>
  </div>
</div>
`;
